aten/src/ATen/native/cpu/Activation.cpp

```diff
@@ -1,31 +1,103 @@
+#include <ATen/Config.h>
 #include <ATen/cpu/vec256/vec256.h>
 #include <ATen/native/TensorIterator.h>
 #include <ATen/native/cpu/Loops.h>
 
-namespace at { namespace native {
+#if AT_MKL_ENABLED()
+#include <mkl.h>
+#endif // AT_MKL_ENABLED()
+
+namespace at {
+namespace native {
+
 namespace {
 
-static void threshold_kernel(TensorIterator& iter, Scalar threshold_scalar, Scalar value_scalar) {
+static void threshold_kernel(
+    TensorIterator& iter,
+    Scalar threshold_scalar,
+    Scalar value_scalar) {
   AT_DISPATCH_ALL_TYPES(iter.dtype(), "threshold_cpu", [&] {
     using Vec = Vec256<scalar_t>;
     scalar_t threshold = threshold_scalar.to<scalar_t>();
     scalar_t value = value_scalar.to<scalar_t>();
     binary_kernel_vec(
-      iter,
-      [&](scalar_t x, scalar_t other) -> scalar_t {
-        return x <= threshold ? value : other;
-      },
-      [&](Vec x, Vec other) -> Vec {
-        return Vec::blendv(other, Vec(value), x <= Vec(threshold));
-      });
+        iter,
+        [&](scalar_t x, scalar_t other) -> scalar_t {
+          return x <= threshold ? value : other;
+        },
+        [&](Vec x, Vec other) -> Vec {
+          return Vec::blendv(other, Vec(value), x <= Vec(threshold));
+        });
   });
 }
 
-} // anonymous namespace
+#if AT_MKL_ENABLED()
+
+template <typename T>
+void GeluKernelMKLImpl(const Tensor& X, Tensor* Y);
+
+#define DELEGATE_GELU_KERNEL_MKL_IMPL(T, CdfNormFunc, MulFunc) \
+  template <>                                                  \
+  void GeluKernelMKLImpl<T>(const Tensor& X, Tensor* Y) {      \
+    const int64_t N = X.numel();                               \
+    const T* X_data = X.data<T>();                             \
+    T* Y_data = Y->data<T>();                                  \
+    CdfNormFunc(N, X_data, Y_data);                            \
+    MulFunc(N, X_data, Y_data, Y_data);                        \
+  }
+DELEGATE_GELU_KERNEL_MKL_IMPL(float, vsCdfNorm, vsMul)
+DELEGATE_GELU_KERNEL_MKL_IMPL(double, vdCdfNorm, vdMul)
+#undef DELEGATE_GELU_KERNEL_MKL_IMPL
+
+#else // AT_MKL_ENABLED()
+
+template <typename T>
+void GeluKernelMKLImpl(const Tensor& X, Tensor* Y) {
+  AT_ASSERTM(false, "ATen not compiled with MKL");
+}
+
+#endif // AT_MKL_ENABLED()
+
+template <typename T>
+void GeluKernelImplInternal(const Tensor& X, Tensor* Y) {
+  const int64_t N = X.numel();
+  const T* X_data = X.data<T>();
+  T* Y_data = Y->data<T>();
+  for (int64_t i = 0; i < N; ++i) {
+    Y_data[i] = X_data[i] * M_SQRT1_2;
+  }
+  Y->erf_();
+  for (int64_t i = 0; i < N; ++i) {
+    Y_data[i] = (Y_data[i] + T(1)) * X_data[i] * T(0.5);
+  }
+}
+
+// TODO(yangxm): Add another fast kernel using formula
+// y = 0.5x * (1 + tanh(sqrt(2/Pi) * (x + 0.044715x^3)))
+// and the fast tanh impl from Eigen.
+void GeluKernelImpl(const Tensor& X, Tensor* Y) {
+  if (at::hasMKL()) {
+    AT_DISPATCH_FLOATING_TYPES(X.scalar_type(), "GeluKernelImpl", [&]() {
+      GeluKernelMKLImpl<scalar_t>(X, Y);
```

Comment:
Would using `TensorIterator` remove the need of using `MKL` functions altogether, and allow for computing everything without having to convert the tensors to contiguous?

Reply:
I guess it will be a little tricky since the reason we try to use MKL is to use the v?CdfNorm function in MKL. So if we want to use TensorIterator here, I guess we have to implement the related things for CdfNorm first?

Reply:
that's not conflicting, as in you can use TensorIterator, where the specialization for a contiguous block can call v?CdfNorm if MKL is available. It's decoupled from using vec256.h.

But just do that in a follow-up diff, instead of this one, as it's an independent work unit anyways.

Reply:
With `TensorIterator`, you could just bypass the need of `v?CdfNorm`, and use instead the methods that we have on `Vec256`, as we have `erf` implemented in `Vec256` as well.

So this would look something like, the following, which is copied from https://github.com/pytorch/pytorch/blob/052bab70697bf6cfe18570374ff26fce93dace94/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L33-L46

```c
static void gelu_kernel(TensorIterator& iter) {
  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "sigmoid_cpu", [&]() {
    unary_kernel_vec(
        iter,
        [=](scalar_t a) -> scalar_t {
          return (std::erf(a * (scalar_t)M_SQRT1_2) + (scalar_t)1) * a * (scalar_t)0.5; },
        [=](Vec256<scalar_t> a) {
          auto b = (Vec256<scalar_t>((scalar_t)(M_SQRT1_e)) * a;
          b = b.erf() + Vec256<scalar_t>((scalar_t)(1));
          return Vec256<scalar_t>((scalar_t)(0.5)) * a * b;
        });
  });
}
```

I'm not sure how much slower it would be compared to using `CdfNorm`, but this implementation seems simpler, handles non-contiguous tensors and uses vectorization thanks to sleef implementation of `erf`.

Or, as @soumith pointed out, one can also use `v?CdfNorm` on contiguous blocks, maybe taking inspiration from https://github.com/pytorch/pytorch/blob/052bab70697bf6cfe18570374ff26fce93dace94/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L190-L221

Reply:
Actually I have tried to use TensorIterator with Vec256. In that case, it was still about 2x slower than the current approach on contiguous case.

Reply:
I'm sorry that I made a mistake. Actually the reason that current approach using erf is faster is Tensor::erf() may also use MKL functions. If I disable MKL, the results are the same. So the conclusion is, if we have MKL, both using CdfNorm or Erf, we can be 2x faster than using Vec256 with MKL. I published https://github.com/pytorch/pytorch/pull/21326 to use TensorIterator. Please take a look. Thanks for the suggestions.